python_modules/dagster/dagster/_core/instance/persist_run_for_production.py

```diff
@@ -0,0 +1,71 @@
+from typing import Mapping, Optional
+
+from dagster._core.definitions.utils import validate_tags
+from dagster._core.host_representation.repository_location import RepositoryLocation
+from dagster._core.host_representation.selector import PipelineSelector
+from dagster._core.instance import DagsterInstance
+from dagster._core.storage.pipeline_run import DagsterRun, DagsterRunStatus
+from dagster._core.utils import make_new_run_id
+
+
+def persist_run_for_production(
+    *,
+    instance: DagsterInstance,
+    repository_location: RepositoryLocation,
+    pipeline_selector: PipelineSelector,
+    run_config: Mapping[str, object],
+    context_specific_tags: Mapping[str, str],
+    explicit_mode: Optional[str],
+) -> DagsterRun:
+    """
+    Creates a run suitable for production using host process (i.e. External*) APIs.
+    This orchestrates necessary interactions with the user process (such as
+    fetching the external pipeline appropriate to the passed subset and
+    compiling the execution plan). This persists a run in instance with the
+    NOT_STARTED state.
+
+    Parameters:
+        instance (DagsterInstance): Instace to execute against
+        repository_location (RepositoryLocation): RepositoryLocation corresponding to user code
+        pipeline_selector (PipelineSelector): Selector that encapsulates subset of pipeline that will be executed
+        run_config (Mapping[str, object]): Run configuration for this run
+        context_specific_tags (Mapping[str, str]): Callsites typically have tags
+            specific to their context (e.g. users can specify ad hoc tags for runs.
+            This set of tags will be merged with the tags on the pipeline itself)
+        explicit_mode Optional[str]: Explicitly override the default mode for the pipeline.
+    """
+    external_pipeline = repository_location.get_external_pipeline(pipeline_selector)
```

Comment:
I believe this results in a gRPC call to the user code server.

In the backfill codepaths (maybe elsewhere too), we launch a set of runs that all use the same subset, and we call `get_external_pipeline` just once to avoid N gRPC calls.

Thoughts on how we could make this function usable in that situation without sacrificing perf? E.g. perhaps we could put an LRU cache on `get_external_pipeline`.

Reply:
lru cache seems like a good idea. I can also press ahead and refactor the backfill codepath before committing this to see what feels best.